### PR TITLE
Add SSH username with private key Credentials

### PR DESCRIPTION
**Pull request by ericcitaire**

### CODE REVIEW

**Comment by casz** [comment]
@laurentribot haven't had a chance to review it.
again you can always take the incremental if you need it asap

File: src/main/resources/com/datapipe/jenkins/vault/credentials/common/VaultSSHUserPrivateKeyImpl/credentials.jelly [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  </f:entry>
+  <f:entry title="${%Private Key Key}" field="privateKeyKey" default="private_key">
+    <f:textbox/>
+  </f:entry>
+  <f:entry title="${%Passphrase Key}" field="passphraseKey" default="passphrase">
+    <f:textbox/>
+  </f:entry>
+  <f:entry name="engineVersion" title="${%K/V Engine Version}" field="engineVersion">
+      <select name="engineVersion">
+          <option value="1" selected="${instance.engineVersion.equals(1)? 'true' : null}">V1</option>
```
Comment (casz) [inline src/main/resources/com/datapipe/jenkins/vault/credentials/common/VaultSSHUserPrivateKeyImpl/credentials.jelly]: Should use an `f:select` instead

File: src/main/resources/com/datapipe/jenkins/vault/credentials/common/VaultUsernamePasswordCredentialImpl/credentials.jelly [hunk 2]
```diff
@@ -12,8 +12,8 @@
   </f:entry>
   <f:entry name="engineVersion" title="${%K/V Engine Version}" field="engineVersion">
       <select name="engineVersion">
-          <option value="1">V1</option>
-          <option value="2">V2</option>
+          <option value="1" selected="${instance.engineVersion.equals(1)? 'true' : null}">V1</option>
```
Comment (casz) [inline src/main/resources/com/datapipe/jenkins/vault/credentials/common/VaultUsernamePasswordCredentialImpl/credentials.jelly]: Should use an `f:select` instead

File: src/main/java/com/datapipe/jenkins/vault/credentials/common/VaultSSHUserPrivateKeyBinding.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        UnbindableDir keyDir = UnbindableDir.create(workspace);
+        FilePath keyFile = keyDir.getDirPath().child("ssh-key-" + privateKeyVariable);
+
+        StringBuilder contents = new StringBuilder();
+        for (String key : sshKey.getPrivateKeys()) {
+            contents.append(key);
+            contents.append('\n');
+        }
+        keyFile.write(contents.toString(), "UTF-8");
+        keyFile.chmod(0400);
```
Comment (casz) [inline src/main/java/com/datapipe/jenkins/vault/credentials/common/VaultSSHUserPrivateKeyBinding.java]: This won't work on Windows. 😓 
Please see https://github.com/jenkinsci/git-client-plugin/pull/447

**Comment by casz** [comment]
Hi @ericcitaire can you please move 9e94592 to a separate PR.
I like the improvement but this is not apart of the work in the PR title.
Also it deserves a separate discussion

**Comment by ericcitaire** [comment]
Thank you for your feedback @casz.

I made all the changes you requested.

I also fixed an issue when retrieving secrets from an agent, which is the case when cloning a Git repository.

As soon as the code looks good for you, I can also implement Secret File and Secret Text credentials the same way.

**Comment by ericcitaire** [comment]
@casz for some reason, the build is broken. It seems to be a network problem rather than a problem with the code itself.

Can you re-launch it please?

**Comment by casz** [comment]
to trigger CI it should be as simple as closing, wait a few seconds and open the pull request again.

**Comment by ericcitaire** [comment]
> Hi @ericcitaire can you please move [9e94592](https://github.com/jenkinsci/hashicorp-vault-plugin/commit/9e9459250911f754c0921044aca977208622d9c4) to a separate PR.
> I like the improvement but this is not apart of the work in the PR title.
> Also it deserves a separate discussion

Ok, I made another PR : #76.

**Comment by ericcitaire** [comment]
Hi @casz. Did you have a chance to take a look at the changes I made?

We are currently using the incremental build [3.1.2-rc283.3434b01d7872](https://repo.jenkins-ci.org/incrementals/com/datapipe/jenkins/plugins/hashicorp-vault-plugin/3.1.2-rc283.3434b01d7872/) which is built on PR #76. And it works like a charm! :)

**Pull request merged**